crates/functional-yew/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+use std::borrow::Borrow;
+use std::cell::RefCell;
+use std::ops::DerefMut;
+use std::rc::Rc;
+use yew::{Component, ComponentLink, Html, Properties};
+
+thread_local! {
+    static CURRENT_HOOK: Rc<RefCell<Option<Rc<RefCell<HookState>>>>> = Rc::new(RefCell::new(None));
+}
+
+struct HookState {
+    counter: usize,
+    render_call: Rc<dyn Fn()>,
+    hooks: Vec<Option<Box<dyn std::any::Any>>>,
+}
+
+pub trait FunctionProvider {
+    type TProps: Properties + PartialEq;
+    fn run(props: &Self::TProps) -> Html;
+}
+
+pub struct FunctionComponent<T: FunctionProvider> {
+    _never: Option<std::marker::PhantomData<T>>,
+    props: T::TProps,
+    hook_state: RefCell<Option<Rc<RefCell<HookState>>>>,
+}
+
+impl<T: 'static> Component for FunctionComponent<T>
+where
+    T: FunctionProvider,
+{
+    type Message = ();
+    type Properties = T::TProps;
+
+    fn create(props: Self::Properties, link: ComponentLink<Self>) -> Self {
+        let render_call = Rc::new(move || link.send_message(()));
+        FunctionComponent {
+            _never: None,
+            props,
+            hook_state: RefCell::new(Some(Rc::new(RefCell::new(HookState {
+                counter: 0,
+                render_call,
+                hooks: vec![],
+            })))),
+        }
+    }
+
+    fn update(&mut self, _: Self::Message) -> bool {
+        true
+    }
+
+    fn change(&mut self, props: Self::Properties) -> bool {
+        let mut props = props;
+        std::mem::swap(&mut self.props, &mut props);
+        props == self.props
+    }
+
+    fn view(&self) -> Html {
+        self.hook_state
+            .try_borrow_mut()
+            .expect("Unexpected concurrent/nested view call")
+            .as_mut()
+            .map(|a| {
+                a.clone()
+                    .try_borrow_mut()
+                    .expect("Unexpected borrow for hook state")
+                    .counter = 0
+            });
+        let previous_hook = CURRENT_HOOK.with(|previous_hook| previous_hook.clone());
+        std::mem::swap(
+            previous_hook
+                .try_borrow_mut()
+                .expect("Previous hook still borrowed")
+                .deref_mut(),
+            self.hook_state.try_borrow_mut().unwrap().deref_mut(),
+        );
+
+        let ret = T::run(&self.props);
+
+        std::mem::swap(
+            previous_hook.borrow_mut().deref_mut(),
+            self.hook_state.borrow_mut().deref_mut(),
+        );
+        return ret;
```

Comment:
This is a variation of your code without the need to clone the hook:

```suggestion
        // Reset hook
        self.hook_state
            .try_borrow_mut()
            .expect("Unexpected concurrent/nested view call")
            .as_mut()
            .unwrap()
            .counter = 0;

        // Load hook
        CURRENT_HOOK.with(|previous_hook| {
            std::mem::swap(
                previous_hook
                    .try_borrow_mut()
                    .expect("Previous hook still borrowed")
                    .deref_mut(),
                self.hook_state.borrow_mut().deref_mut(),
            );
        });

        let ret = T::run(&self.props);

        // Unload hook
        CURRENT_HOOK.with(|previous_hook| {
            std::mem::swap(
                previous_hook
                    .try_borrow_mut()
                    .expect("Previous hook still borrowed")
                    .deref_mut(),
                self.hook_state.borrow_mut().deref_mut(),
            );
        });

        return ret;
```

Reply:
More elegant :)
Integrated